### PR TITLE
Controller: Fix BrokerRegistration + BrokerHeartbeat RPC

### PR DESCRIPTION
**Pull request by aloknnikhil**
### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/ControllerApis.scala [hunk 1]
```diff
@@ -220,12 +221,64 @@ class ControllerApis(val requestChannel: RequestChannel,
...
     val registrationRequest = request.body[BrokerRegistrationRequest]
     apisUtils.authorizeClusterOperation(request, CLUSTER_ACTION)
-    controller.registerBroker(registrationRequest.data())
+
+    controller.registerBroker(registrationRequest.data).handle[Unit]((reply, ex) => {
+      var err = Errors.NONE
+      if (ex != null) {
+        error(s"Failed to register broker: ${ex.getMessage}")
+        err = Errors.forException(ex)
+      }
```
Comment (rondagostino) [inline core/src/main/scala/kafka/server/ControllerApis.scala:230]: Change to a `val`?  Same with below.

**Review by rondagostino — COMMENTED** [review]
LGTM, left 1 comment.

File: core/src/main/scala/kafka/server/ControllerApis.scala [hunk 2]
```diff
@@ -220,12 +221,64 @@ class ControllerApis(val requestChannel: RequestChannel,
...
+        val brokerRegistrationResponseData = new BrokerRegistrationResponseData()
+
+        // TODO: get a better estimate of the current active controller from Raft or somewhere.
+        if (controller.curClaimEpoch() > 0) {
+          brokerRegistrationResponseData.setActiveControllerId(config.controllerId)
+        } else {
+          brokerRegistrationResponseData.setActiveControllerId(MetadataResponse.NO_CONTROLLER_ID)
+        }
+        brokerRegistrationResponseData.setBrokerEpoch(reply.epoch)
+        brokerRegistrationResponseData.setLeaseDurationMs(reply.epoch)
```
Comment (cmccabe) [inline core/src/main/scala/kafka/server/ControllerApis.scala:242]: lease duration shouldn't be set as epoch

File: core/src/main/scala/kafka/server/ControllerApis.scala [hunk 3]
```diff
@@ -220,12 +221,64 @@ class ControllerApis(val requestChannel: RequestChannel,
   def handleBrokerRegistration(request: RequestChannel.Request): Unit = {
     val registrationRequest = request.body[BrokerRegistrationRequest]
     apisUtils.authorizeClusterOperation(request, CLUSTER_ACTION)
-    controller.registerBroker(registrationRequest.data())
+
+    controller.registerBroker(registrationRequest.data).handle[Unit]((reply, ex) => {
+      var err = Errors.NONE
```
Comment (cmccabe) [inline core/src/main/scala/kafka/server/ControllerApis.scala:226]: if statements can return a value in scala, so you don't need to mutate this

File: core/src/main/scala/kafka/server/ControllerApis.scala (hunk 2 above)
Comment (aloknnikhil) [reply core/src/main/scala/kafka/server/ControllerApis.scala:242]: Good catch!

File: core/src/main/scala/kafka/server/ControllerApis.scala (hunk 1 above)
Comment (aloknnikhil) [reply core/src/main/scala/kafka/server/ControllerApis.scala:230]: Done.

File: core/src/main/scala/kafka/server/ControllerApis.scala (hunk 3 above)
Comment (aloknnikhil) [reply core/src/main/scala/kafka/server/ControllerApis.scala:226]: Nice! Fixed.

**Comment by cmccabe** [comment]
LGTM

**Pull request merged**